python/lsst/summit/utils/peekExposure.py

```diff
@@ -0,0 +1,772 @@
+
+    def getDonutDiameter(self, exposure):
+        """Estimate donut diameter from exposure metadata.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure to estimate donut diameter for.
+
+        Returns
+        -------
+        donutDiameter : float
+            Estimated donut diameter in pixels.
+        """
+        visitInfo = exposure.getInfo().getVisitInfo()
+        focusZ = visitInfo.focusZ
+        instrumentLabel = visitInfo.instrumentLabel
+
+        match instrumentLabel:
+            case "LATISS":
+                focusZ *= 41  # magnification factor
+                fratio = 18.0
+            case "LSSTCam" | "ComCam":
+                fratio = 1.234
+        # AuxTel/ComCam/LSSTCam all have 10 micron pixels (= 10e-3 mm)
+        donutDiameter = abs(focusZ) / fratio / 10e-3
+        self.log.info(f"{focusZ=} mm")
+        self.log.info(f"donutDiameter = {donutDiameter} pixels")
+        return donutDiameter
+
+    def run(
+        self,
+        exposure,
+        *,
+        doDisplay=False,
+        doDisplayIndices=False,
+        mode="auto",
+        donutDiameter=None,
+    ):
+        """
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure at which to peek.
+        doDisplay : bool, optional
+            Display the exposure and sources?  Default False.
+        doDisplayIndices : bool, optional
+            Display the source indices?  Default False.
+        mode : {'auto', 'donut', 'spec', 'photo'}, optional
+            Mode to run in.  Default 'auto'.
+        donutDiameter : float, optional
+            Donut diameter in pixels.  If None, will be estimated from exposure
+            metadata.  Default None.
+
+        Returns
+        -------
+        result : `pipeBase.Struct`
+            Result of the peek.
+            Struct containing:
+                - mode : str
+                    Peek mode that was run.
+                - binSize : int
+                    Binning factor used.
+                - binnedSourceCat : `lsst.afw.table.SourceCatalog`
+                    Source catalog from the binned exposure.
+                - table : `astropy.table.Table`
+                    Curated source table in unbinned coordinates.
+                - brightestIdx : int
+                    Index of the brightest source in the source catalog.
+                - brightestCentroid : `lsst.geom.Point2D`
+                    Centroid of the brightest source (unbinned coords).
+                - brightestShape : `lsst.geom.Quadrupole`
+                    Shape of the brightest source (unbinned coords).
+                - psfShape : `lsst.geom.Quadrupole`
+                    Estimated PSF shape (unbinned coords).
+        """
+        # We're going to do binning and other manipulations, so just make a
+        # copy now and don't worry about modifying the original.
+        exposure = exposure.clone()
```

Comment:
There's no way binning is going to preserve the original image, so if you can make sure that happens first this deep-copy won't be necessary, and that might be useful if you're trying to squeeze performance.